common/json.c

```diff
@@ -493,6 +442,56 @@ void json_add_string(struct json_result *result, const char *fieldname, const ch
 	result_append_fmt(result, "\"%s\"", escaped);
 }
 
+void json_add_string_escape(struct json_result *result, const char *fieldname,
+			    const char *value)
+{
+	/* Worst case: all \uXXXX */
+	char *escaped = tal_arr(result, char, strlen(value) * 6 + 1);
+	size_t i, n;
+
+	json_start_member(result, fieldname);
+	for (i = n = 0; value[i]; i++, n++) {
+		char esc = 0;
+		switch (value[i]) {
+		case '\n':
+			esc = 'n';
+			break;
+		case '\b':
+			esc = 'b';
+			break;
+		case '\f':
+			esc = 'f';
+			break;
+		case '\t':
+			esc = 't';
+			break;
+		case '\r':
+			esc = 'r';
+			break;
+		case '\\':
+		case '"':
+			esc = value[i];
+			break;
+		default:
+			if ((unsigned)value[i] < ' '
```

Comment:
How about bytes 0x80 and above? `char` may be unsigned on some platforms. Better to also add a `|| value[i] & 0x80` condition.

Reply:
No, ignore this comment.  Bytes 0x80 and above are UTF-8 multi-byte codes, and JSON allows Unicode code points above the ASCII range to be encoded directly without an escape, and JSON itself should be encoded in UTF-8 by default (and since we use primarily ASCII, we implicitly use UTF-8 by default).  Regardless of signedness of `char` this should still let 0x80 and above to pass through (due to `(unsigned)` conversion).  There is the issue of invalid UTF-8 being generated but all our messages are ASCII, and any non-ASCII text, will be from the user anyway (so it is user fault if user gave non-UTF-8 to us that we are simply returning back to the user).